src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,296 @@
+//
+// Author: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+
+#include <sof/list.h>
+#include <sof/spinlock.h>
+#include <sof/audio/component.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+/* per-scheduler data */
+struct zephyr_ll {
+	struct list_item tasks;			/* list of ll tasks */
+	spinlock_t lock;			/* protects the list or tasks */
+	struct ll_schedule_domain *ll_domain;	/* scheduling domain */
+};
+
+/* per-task scheduler data: so far actually unused */
+struct zephyr_ll_pdata {
+	uint64_t period;
+};
+
+//FIXME
+#define domain_time(sch) platform_timer_get_atomic(timer_get())
+
+static void zephyr_ll_run(void *data)
+{
+	struct zephyr_ll *sch = data;
+	struct task *task;
+	struct list_item *list, *tmp;
+
+// FIXME: locking
+	/*
+	 * We cannot use a mutex because tasks can be added or deleted from
+	 * interrupt context. TODO: use two lists: staging and running. Use the
+	 * lock to access the staging list, only access the running list in this
+	 * function - with no locking. To cancel set the state to CANCEL, here
+	 * check before running with lock held. Then release the lock and run
+	 * the task. If the task is cancelled during running - that's normal.
+	 * After task return take the lock, if cancelled - remove from the list
+	 * and set to FREE.
+	 */
+	spin_lock(&sch->lock);
+	list_for_item_safe(list, tmp, &sch->tasks) {
+		struct comp_dev *sched_comp;
+		enum task_state state;
+		uint32_t flags;
+
+		task = container_of(list, struct task, list);
+
+		if (task->state == SOF_TASK_STATE_CANCEL) {
+			list_item_del(&task->list);
+			task->state = SOF_TASK_STATE_FREE;
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
+			continue;
+
+		task->state = SOF_TASK_STATE_RUNNING;
+		state = task_run(task);
+
+		if (state == SOF_TASK_STATE_COMPLETED)
+			list_item_del(&task->list);
+
+		if (task->state == SOF_TASK_STATE_RUNNING)
+			task->state = state;
+
+		task->start = sch->ll_domain->next_tick;
+	}
+	spin_unlock(&sch->lock);
+}
+
+/* Called once for periodic tasks or multiple times for one-shot tasks */
+static int zephyr_ll_task_schedule(void *data, struct task *task, uint64_t start,
+				   uint64_t period)
+{
+	struct zephyr_ll *sch = data;
+	struct zephyr_ll_pdata *pdata;
+	struct task *task_iter;
+	struct list_item *list;
+	uint64_t delay = period ? period : start;
+	int ret;
+
+	spin_lock(&sch->lock);
+
+// FIXME: locking
+	/* check if task is already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock(&sch->lock);
+			return 0;
+		}
+	}
+
+	pdata = task->priv_data;
+	pdata->period = period;
+
+	task->start = domain_time(sch) + delay;
+	if (sch->ll_domain->next_tick != UINT64_MAX &&
+	    sch->ll_domain->next_tick > task->start)
+		task->start = sch->ll_domain->next_tick;
```

Comment:
I dont think we need to have the concept of task start time here as the task will only start at the next LL tick iff it's in the task list.

Reply:
let's look at this once it's first working reliably as is, I'll want to understand this better.

Reply:
I assume this is removed in the subsequent PR ?

Reply:
it will be, yes, adding comments for that